text/0000-min-target-api-version.md

```diff
@@ -0,0 +1,205 @@
+
+Implementors will likely want to implement `Ord`, `FromStr` and `Dispay` on some enum which represents a given platform's api version numbers. 
+
+### Storing the information in metadata
+
+The values for the various `cfg` predicates must be stored in a crate's metadata if that value is set. If none is set than the value is defaulted to whatever the maximum of all that crate's transitive dependencies is.
+
+### Future compatibility
+
+A given version of the Rust compiler will only know about certain versions of a given API version string. If a compiler does not recognize the given version string, it will issue a warning and the `cfg` predicate will return false. This means as new versions of platforms are introduced, new support must be added to the compiler. As code is written to take advantage of these new version strings, older compilers will not be able to conditionally compile based on them.  
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+There are no known large drawbacks to this proposal. Some small drawbacks include:
+* No all targets will strictly follow an ever increasing versioning scheme where more recent versions (i.e., versions with larger version numbers) are supersets of less recent versions. APIs may be deprecated and completely removed between versions. 
+   * This proposal does not seek to address these situations. The mechanism does not provide any semantic versioning scheme on the versioning numbers. Maintainers must ensure that any feature they use can work for _all_ versions greater than or equal to the target API version specified.
+   * See the [future possibilities](#future-posibilities) section for a possible `max_*` variant which can address these situations.
+   * This does mean that code compiled with minimum target API versions is assumed to continue to work with all future versions of a target API. This is not considered a drawback per say since this is currently implicitly true for any target API use. 
+* There is the potential for an ever growing number of built-in `cfg` predicates. As support is added for other platforms (e.g., iOS, Android, BSD, etc.), there will be a need for new `cfg` predicates that fit those platforms.
+* Incremental complexity of the language
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The overall mechanism proposed here builds on other well established primitives in Rust such as `cfg`. 
+
+As previously stated, a mechanism which tries to bridge cross-platform differences under one `min_target_api_version` predicate [was suggested](https://github.com/rust-lang/rfcs/blob/b0f94000a3ddbd159013e100e48cd887ba2a0b54/text/0000-min-target-api-version.md) but was rejected due to different platforms having divergent needs.
+
+To combat the issue of ever increasing number of platform API `cfg` predicates, a new syntax could be introduce like `#[cfg(min_version("windows_build", "6.0.6000")]`
```

Comment:
I would strongly prefer this. Firstly this avoids the can of worms related to backwards compatibility. 

As a secondhand benefit something like this could also pave the road to different matching predicates within the version string, much like what we have in cargo. So for instance, crates use semver and so a variety of operators that make sense for semver can be used (rather than just `>=`):

```
library = "^1.0"
other_library = ">= 1.0 && <3.34"
```

If, out there, there's a target that utilizes semantic versioning for itself or its components (I think `freebsd` is a possible close match here), then utilizing the familiar semver matching algorithms would also make sense for it.

Reply:
Specifying a generic mechanism for targets to provide their own versioning schemes, rather than concrete `cfg`s also sounds like a more productive direction to me. Then the fate of this RFC wouldn't need to linger on us being comfortable with all sorts of versioning schemes proposed for the various targets. Once an RFC with a generic mechanism is merged, we could spend our time on figuring out out the high value versioning cfgs (windows, macos) sooner via other channels (PR FCPs, MCPs etc) even before we know what to do about linux, android or bsds.

Reply:
Amen.  Better to have a single syntax for all operating systems, rather than defining a separate variable name for each.